oeps/oep-0026-arch-xapi-integration.rst

```diff
@@ -0,0 +1,483 @@
+
+The **Context** field allows us to embed additional contextual information in each statement.
+See specifics in the `Open edX Events`_ section below since it varies by event type.
+
+Here is an example of a **Context** JSON value that we would generate:
+
+::
+
+    "context": {
+        "registration": "openedx.org/enrollments/enrollment-v1:<anonymized-enrollment-id>",
+        "contextActivities": {
+            “parent”: [{
+                “objectType”: “Activity”,
+                “id”: “https://openedx.org/courses/course-v1:openedx+origami-folding+1T2018"
+            }]
+        }
+    }
+
+See `Deep Dive: Context`_ for more information on xAPI Context.
+
+.. _`Deep Dive: Context`: https://xapi.com/deep-dive-context/
+
+xAPI Result
+~~~~~~~~~~~
+
+The **Result** field specifies the score the user earned on an activity.  Here is an
+example of a JSON value that we would generate for a problem type:
+
+::
+
+    "result": {
+        "success": false,
+        "completion": true,
+        "score": {
+            "min": 0,
+            "max": 50,
+            "raw": 10,
+            "scaled": .20
+        }
+        "response": "foo"
+    }
+
+See `Deep Dive: Result`_ for more information on xAPI Result.
+
+.. _`Deep Dive: Result`: https://xapi.com/deep-dive-result/
+
+Open edX xAPI Plugin
+====================
+
+A new library will be implemented, using the Open edX `Django App Plugin`_ design pattern,
+that:
+
+1. Translates Open edX native events to Open edX xAPI events (**Translator**).
+2. Validates the generated event using a general-purpose validation library 
+   cross-maintained by Open edX xAPI consumers (**Validator**).
+3. Routes real-time xAPI events to all registered consumers (**Router**).
+4. Provides an administration interface to configure and manage registered
+   consumers (**Admin**).
+
+Translator
+~~~~~~~~~~
+
+This component will plug into the `Open edX Event Tracking`_ subsystem to listen to
+events of interest and translate them to xAPI's JSON format.
+
+TBD - The development team will assess whether we should use (and start owning)
+the already implemented (but no longer maintained) `xAPI Python Open Source Library`_. 
+
+Validator
+~~~~~~~~~
+
+The intention of this validation component is to ensure we continue to support the
+expected xAPI formats by all participating Open edX xAPI consumers. In many ways,
+this is similar to what `consumer-driven contract testing`_ would enable and uses
+similar design principles. Essentially, a common validation library can be
+collaboratively maintained by participating consumers, including consuming adaptive
+engines. This ensures an end-to-end integration that can be maintained going forward.
+
+**Note**: Although xAPI specifies a standardized format, it is a low-level transaction
+schema and relies on higher-level "profiles" applied on top of it. So the profiles
+for specific Activities, Verbs, Contexts, etc used by Open edX need to be contractually
+maintained.
+ 
+Router
```

Comment:
What queue are you going to use?
maybe the first consumer of the queue is the LRS.

Reply:
Please see the updated version of the spec - as this has changed thinking about scalability concerns.